GeometryCore/src/LinearAlgebra.cpp

```diff
@@ -4,29 +4,114 @@
+        {
+            return true;
+        }
+        return false;
     }
 
     // Return distance from the p to the line
-    FPoint2D::coord distanceFromPointToLine(const FPoint2D& p, const LineSegment2D& line)
-    {
-        return {};
+    FPoint2D::coord distanceFromPointToLine(const FPoint2D& p, const LineSegment2D& line) {
+        auto projection = pointProjectionToLineSegment(p, line);
+
+        if ((projection.x >= std::min(line.a.x, line.b.x)) && (projection.x <= std::max(line.a.x, line.b.x)) &&
+            (projection.y >= std::min(line.a.y, line.b.y)) && (projection.y <= std::max(line.a.y, line.b.y))) {
+            return distance(projection, p);
+        }
+        else {
+            return std::min(distance(line.a, p), distance(line.b, p));
+        }
+    }
+
+    FPoint2D::coord distanceBetweenLineSegments(const LineSegment2D& line1, const LineSegment2D& line2) {
+        auto check_projection = [](const FPoint2D& p, const LineSegment2D& line, FPoint2D::coord& minimal_distance) { minimal_distance = std::min(distanceFromPointToLine(p, line), minimal_distance); };
+
+        FPoint2D::coord min_dis = std::min(std::min(distance(line1.a, line2.a), distance(line1.a, line2.b)), std::min(distance(line1.b, line2.a), distance(line1.b, line2.b)));
+        check_projection(line2.a, line1, min_dis);
+        check_projection(line2.b, line1, min_dis);
+        check_projection(line1.a, line2, min_dis);
+        check_projection(line1.b, line2, min_dis);
+
+        return min_dis;
     }
 
     // Return true if poly and line are colliding with each other.
     // Note: consider collision even if one of line end lies exactly on the polygon contour
-    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line)
-    {
-        return {};
+    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line) {
+        auto convex = poly.convexHull();
+        std::vector<bool> line_collide_with_polygon_side;
+        for (int i = 0; i < convex.size(); ++i) {
+                LineSegment2D lie(convex[i], convex[(i + 1) % convex.size()]);
+                line_collide_with_polygon_side.push_back(lineSegmentsCollide(line, lie));
+        }
+        if (std::any_of(line_collide_with_polygon_side.begin(), line_collide_with_polygon_side.end(), [](bool k) {return k; })) {
+            return true;
+        }
+        return false;
+    }
+
+    //Return true if polygon is close to line (closer than given gap)
+    bool polygonAndLineSegmentAreAdjacent(const Polygon& poly, const LineSegment2D& line, const FPoint2D::coord gap) {
+        auto convex = poly.convexHull();
+        std::vector<FPoint2D::coord> distance_from_line_to_polygon_side;
+        #pragma omp parallel for
+        for (int i = 0; i < convex.size(); ++i) {
+                LineSegment2D lie(convex[i], convex[(i + 1) % convex.size()]);
+                #pragma omp atomic
```

Comment:
Ok. Or you can resize vector to hold enough amount of elements before parallel loop (size should be equal poly.size()) to assign distanceBetweenLineSegments() value for each element in a vector. Thus, you can remove `#pragma omp atomic` to eliminate additional overhead due to synchronization and let simultaneous access to vector.

Reply:
Yes, this variant will be better, I will implement this